libcudacxx/include/cuda/std/__complex/inverse_hyperbolic_functions.h

```diff
@@ -21,69 +21,352 @@
 #  pragma system_header
 #endif // no system header
 
-#include <cuda/std/__cmath/abs.h>
-#include <cuda/std/__cmath/copysign.h>
-#include <cuda/std/__cmath/isinf.h>
-#include <cuda/std/__cmath/isnan.h>
-#include <cuda/std/__cmath/trigonometric_functions.h>
-#include <cuda/std/__complex/complex.h>
-#include <cuda/std/__complex/exponential_functions.h>
-#include <cuda/std/__complex/logarithms.h>
 #include <cuda/std/__complex/nvbf16.h>
 #include <cuda/std/__complex/nvfp16.h>
 #include <cuda/std/__complex/roots.h>
-#include <cuda/std/limits>
-#include <cuda/std/numbers>
 
 #include <cuda/std/__cccl/prologue.h>
 
 _CCCL_BEGIN_NAMESPACE_CUDA_STD
 
+// Specialization of 1/sqrt(x), uses device-only rsqrtf/rsqrt when viable.
+// Sometimes compilers can optimize 1/sqrt(x) so we don't suffer the slowdown of the divide.
+// It can happen that the double 1/sqrt() on device gets optimized better than CUDA rsqrt(),
+// but it depends on the calling function. We needs to optimize on an individual basis.
+template <class _Tp>
+[[nodiscard]] _CCCL_API inline _Tp __internal_rsqrt_inverse_hyperbloic(_Tp __x) noexcept
+{
+  if constexpr (is_same_v<_Tp, float>)
+  {
+    NV_IF_TARGET(NV_IS_DEVICE, (return ::rsqrtf(__x);))
+  }
+  if constexpr (is_same_v<_Tp, double>)
+  {
+    NV_IF_TARGET(NV_IS_DEVICE, (return ::rsqrt(__x);))
+  }
+  return _Tp(1) / ::cuda::std::sqrt(__x);
+}
+
+// An unsafe sqrt(_Tp + _Tp) extended precision sqrt.
+template <typename _Tp>
+static void __device__ __host__ __forceinline__
+__internal_double_Tp_sqrt_unsafe(_Tp __hi, _Tp __lo, _Tp* __out_hi, _Tp* __out_lo) noexcept
+{
+  // rsqrt
+  const _Tp __initial_guess = __internal_rsqrt_inverse_hyperbloic<_Tp>(__hi);
+
+  // Newton-Raphson, assume we have a reasonable rsqrt above and
+  // that we don't need to update the first term (__initial_guess).
+  //     x_(n+1) = x_n - 0.5*x_n*((__hi + __lo)(x_n^2) - 1)
+
+  // __initial_guess^2:
+  const _Tp __init_sq_hi = __initial_guess * __initial_guess;
+  const _Tp __init_sq_lo = ::cuda::std::fma(__initial_guess, __initial_guess, -__init_sq_hi);
+
+  // Times (__hi + __lo).
+  // We need to add the -1.0 here in an fma, or different compilers (eg host vs device)
+  // optimize differently and give (sometimes very) different results.
+  const _Tp _hi_hi_hi = ::cuda::std::fma(__hi, __init_sq_hi, -1.0);
```

Comment:
```suggestion
  const _Tp _hi_hi_hi = ::cuda::std::fma(__hi, __init_sq_hi, _Tp(-1.0));
```
otherwise we would get a conversion to `double`